File: bench/c/good/loops.c

void loop1(int n) {
  //@ variant n
  while (n > 0) n--;
}

void loop2(int n) {
  //@ variant 100-n
  while (n < 100) n++;
}

//@ ensures \result == 0
int loop3() {
  int i = 100;
  //@ invariant 0 <= i variant i
  while (i > 0) i--;
  return i;
}

//@ ensures \result == 100
int sum() {
  int i = 100, s;
  //@ invariant 0 <= i <= 100 && s == 100-i variant i
  for (s = 0; i > 0; i--) s++;
  return s;
} 


//@ requires 0 <= n
void loop(int n) {
  int i;
  //@ invariant 0 <= i <= n variant n-i
  for (i = 0; i < n; i++);
}

File: bench/good/loops.ml

(** 1. A loop increasing [i] up to 10. *)

parameter i : int ref

let loop1 = 
  { i <= 10 }
  while !i < 10 do
    { invariant i <= 10 variant 10-i }
    i := !i + 1
  done
  { i = 10 }


(** 2. The same loop, followed by a function call. *)

parameter x: int ref

let oppose = fun (u:unit) -> (x := - !x) { x = -x@ }

let loop2 = 
  { x <= 10 }
  begin
    while !x < 10 do { invariant x <= 10 variant 10-x } x := !x + 1 done
    { x = 10 };
    (if !x > 0 then (oppose void)) 
    { x = -10 }
  end
